src/targeting.js

```diff
@@ -181,6 +181,48 @@ export function newTargeting(auctionManager) {
     return [];
   };
 
+  /**
+   * Returns filtered ad server targeting for custom and allowed keys.
+   * @param {targetingArray} targeting
+   * @param {string[]} allowedKeys
+   * @return {targetingArray} filtered targeting
+   */
+  function getAllowedTargetingKeyValues(targeting, allowedKeys) {
+    logInfo(`Allowed default targeting keys [ ${allowedKeys.join(', ')}]`);
+    const defaultKeyring = Object.assign({}, CONSTANTS.TARGETING_KEYS, CONSTANTS.NATIVE_KEYS);
+    const defaultKeys = Object.keys(defaultKeyring);
+    targeting.map(adUnit => {
+      const adUnitCode = Object.keys(adUnit)[0];
+      const keyring = adUnit[adUnitCode];
+      const keys = keyring.filter(kvPair => {
+        const key = Object.keys(kvPair)[0];
+        // check if key is in default keys, if not, it's custom, we won't remove it.
+        const isCustom = defaultKeys.filter(defaultKey => key.indexOf(defaultKeyring[defaultKey]) === 0).length === 0;
+        // check if key explicitly allowed, if not, we'll remove it.
+        const found = isCustom || allowedKeys.find(allowedKey => {
+          const allowedKeyName = defaultKeyring[allowedKey];
+          // we're looking to see if the key exactly starts with one of our default keys.
+          // (which hopefully means it's not custom)
+          const found = key.indexOf(allowedKeyName) === 0;
+          return found;
+        });
+        const keyDisposition = found ? 'Keeping' : 'Removing';
+        const keyType = isCustom ? 'custom' : 'default';
+        logInfo(`${keyDisposition} ${keyType} targeting key ${key} for ${adUnitCode} adserverTargeting object`);
+        return found;
+      });
+      adUnit[adUnitCode] = keys;
+    });
+    const filteredTargeting = targeting.filter(adUnit => {
+      const adUnitCode = Object.keys(adUnit)[0];
+      const keyring = adUnit[adUnitCode];
+      const targetingObjectDisposition = keyring.length > 0 ? 'Keeping populated' : 'Removing empty';
+      logInfo(`${targetingObjectDisposition} ${adUnitCode} adserverTargeting object`);
```

Comment:
After running a test I'm not sure if we'd need this log? Or maybe it could possibly provide more info (like what is it keeping). When I ran your code I would see multiple logs either saying "keeping populated" or "removing empty" for the same adunit but it doesn't say much else and we're already logging all the keys removed and kept. Maybe I'm misunderstanding the needs for this as well.

Reply:
This log was intended to illustrate that targeting for the specified adUnit was completely removed and thusly the adUnit was removed from targeting entirely.

Maybe removing the adUnit for lack of targeting is the wrong move in the first place. I'll check the behavior in the case where all targeting is eliminated, but the adUnit isn't removed from the targeting object.

Reply:
After testing the behavior again, I see what you mean, it's neither descriptive nor necessary to log at this step.

Reply:
This has been updated.